ui/src/TestUserInput.ts

```diff
@@ -3,52 +3,51 @@
-    public async showQuickPick(items: Pick[] | Thenable<Pick[]>, placeHolder: string, ignoreFocusOut?: boolean): Promise<Pick>;
-    public async showQuickPick(items: vscode.QuickPickItem[] | Thenable<vscode.QuickPickItem[]>, placeHolder: string, _ignoreFocusOut: boolean = false): Promise<vscode.QuickPickItem> {
+    public async showQuickPick<T extends QuickPickItem>(items: T[] | Thenable<T[]>, options: QuickPickOptions): Promise<T> {
         if (this._inputs.length > 0) {
             const input: string | undefined = this._inputs.shift();
-            const resolvedItems: vscode.QuickPickItem[] = await Promise.resolve(items);
+            const resolvedItems: T[] = await Promise.resolve(items);
 
             if (resolvedItems.length === 0) {
-                throw new Error(`No quick pick items found. Placeholder: '${placeHolder}'`);
+                throw new Error(`No quick pick items found. Placeholder: '${options.placeHolder}'`);
             } else if (input) {
-                const resolvedItem: vscode.QuickPickItem | undefined = resolvedItems.find((qpi: vscode.QuickPickItem) => qpi.label === input);
+                const resolvedItem: T | undefined = resolvedItems.find((qpi: T) => qpi.label === input || qpi.description === input);
                 if (resolvedItem) {
                     return resolvedItem;
                 } else {
-                    throw new Error(`Did not find quick pick item matching '${input}'. Placeholder: '${placeHolder}'`);
+                    throw new Error(`Did not find quick pick item matching '${input}'. Placeholder: '${options.placeHolder}'`);
                 }
             } else {
                 // Use default value if input is undefined
                 return resolvedItems[0];
             }
         }
 
-        throw new Error(`Unexpected call to showQuickPick. Placeholder: '${placeHolder}'`);
+        throw new Error(`Unexpected call to showQuickPick. Placeholder: '${options.placeHolder}'`);
     }
 
-    public async showInputBox(placeHolder: string, prompt: string, _ignoreFocusOut: boolean = false, validateInput?: (s: string) => string | undefined | null, value?: string): Promise<string> {
+    public async showInputBox(options: InputBoxOptions): Promise<string> {
         if (this._inputs.length > 0) {
             let result: string | undefined = this._inputs.shift();
             if (!result) {
                 // Use default value if input is undefined
-                result = value;
+                result = options.value;
             }
 
             if (result) {
-                if (validateInput) {
-                    const msg: string | null | undefined = validateInput(result);
+                if (options.validateInput) {
+                    const msg: string | null | undefined = await Promise.resolve(options.validateInput(result));
```

Comment:
await will wait on both the inner and outer promises if validateInput is a promise?

Reply:
Promise.resolve will just return the original promise if the argument is already a promise. There will be no "inner" and "outer" promise in that case. From [the docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve):


> The Promise.resolve(value) method returns a Promise object that is resolved with the given value. If the value is a thenable (i.e. has a "then" method), the returned promise will "follow" that thenable, adopting its eventual state; if the value was a promise, that object becomes the result of the call to Promise.resolve; otherwise the returned promise will be fulfilled with the value.